webapp/channels/src/components/browse_channels/browse_channels.tsx

```diff
@@ -212,13 +219,27 @@ export default class BrowseChannels extends React.PureComponent<Props, State> {
     };
 
     setSearchResults = (channels: Channel[]) => {
-        this.setState({searchedChannels: this.state.shouldShowArchivedChannels ? channels.filter((c) => c.delete_at !== 0) : channels.filter((c) => c.delete_at === 0), searching: false});
+        // filter out private channels that the user is not a member of
+        let searchedChannels = channels.filter((c) => c.type !== Constants.PRIVATE_CHANNEL || this.isMemberOfChannel(c.id));
+        if (this.state.filter === FILTER.private) {
+            searchedChannels = channels.filter((c) => c.type === Constants.PRIVATE_CHANNEL && this.isMemberOfChannel(c.id));
+        }
+        if (this.state.filter === FILTER.public) {
+            searchedChannels = channels.filter((c) => c.type === Constants.OPEN_CHANNEL && c.delete_at === 0);
+        }
+        if (this.state.filter === FILTER.archived) {
+            searchedChannels = channels.filter((c) => c.delete_at !== 0);
+        }
+        if (this.props.shouldHideJoinedChannels) {
+            searchedChannels = this.getChannelsWithoutJoined(searchedChannels);
+        }
```

Comment:
nit : can this be if else ladder the last else being https://github.com/mattermost/mattermost/pull/24099/files#diff-b0b14d135dc7be9da8fb91320dd550821ed8ef4c29393a1bfbd4ccdb9cfa3dd2R223